brave/src/main/java/brave/sampler/RateLimitingSampler.java

```diff
@@ -0,0 +1,58 @@
+package brave.sampler;
+
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Based on https://github.com/aws/aws-xray-sdk-java/blob/2.0.1/aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/strategy/sampling/reservoir/Reservoir.java
+ *
+ * Uses only Java 6 APIs to be compatible with Brave Core requirements. Accounts for nanoTime()
+ * overflow, as unlikely as that might be.
+ */
+public class RateLimitingSampler extends Sampler {
+  public static Sampler create(int tracesPerSecond) {
+    if (tracesPerSecond <= 0) {
+      return Sampler.NEVER_SAMPLE;
+    }
+    return new RateLimitingSampler(tracesPerSecond);
+  }
+
+  private static long NANOS_PER_SECOND = 1_000_000_000;
+
+  private int tracesPerSecond;
+  private AtomicLong usage;
+  private AtomicLong nextUpdate;
+
+  RateLimitingSampler(int tracesPerSecond) {
+    this.tracesPerSecond = tracesPerSecond;
+    this.usage = new AtomicLong(0);
+    this.nextUpdate = new AtomicLong(getNextUpdateValue(System.nanoTime()));
+  }
+
+
+  @Override public boolean isSampled(long ignoredTraceId) {
+    long now = System.nanoTime();
+    long updateAt = nextUpdate.get();
+
+    // now is past update time OR
+    //      (update time is positive (near Long.MAX_VALUE) AND now is negative (long overflow))
+    if (now > updateAt || (updateAt > 0 && now < 0)) {
+      if (nextUpdate.compareAndSet(updateAt, getNextUpdateValue(updateAt))) {
+        usage.set(1);
```

Comment:
Would use 0 and `incrementAndGet` since it's a bit weird for empty to have a non-zero value.